rpc/transport/ws/server.go

```diff
@@ -71,18 +72,60 @@ func (wsc *serverWebSocketTransport) Close() {
 }
 
 func (wsc *serverWebSocketTransport) Url() string {
-	return "ws://" + webscocketServerAddress + wsc.port
+	return webscocketServerAddress + wsc.port
+}
+
+func (wsc *serverWebSocketTransport) request(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+		return
+	}
+	body := http.MaxBytesReader(w, r.Body, 8192)
+	msg, err := io.ReadAll(body)
+	if err != nil {
+		http.Error(w, http.StatusText(http.StatusRequestEntityTooLarge), http.StatusRequestEntityTooLarge)
+		return
+	}
+
+	_, err = w.Write(wsc.requestHandler(msg))
+	if err != nil {
+		panic(err)
+	}
 }
 
-func (wsc *serverWebSocketTransport) subscribeRequestHandler(w http.ResponseWriter, r *http.Request) {
+func (wsc *serverWebSocketTransport) subscribe(w http.ResponseWriter, r *http.Request) {
 	c, err := websocket.Accept(w, r, nil)
 	if err != nil {
 		panic(err)
 	}
-	wsc.serverWebsocket = c
 	defer c.Close(websocket.StatusInternalError, "server initiated websocket close")
+	notificationChan := make(chan []byte)
+	key := strconv.Itoa(int(rand.Uint64()))
+	wsc.notificationListeners.Store(key, notificationChan)
+	defer wsc.notificationListeners.Delete(key)
+
+	// A client closes a connection by sending a message over the websocket
+	closeChan := make(chan error)
+	go func() {
```

Comment:
Is this meant to block the server until we receive a close signal from the client? It looks like we expect an `err` from `c.Read`, is that because we expect some kind of closed websocket error (caused by the client closing their ws)? 

Should there be a way for the server to close without waiting on the client? Like a `select` statement that also checks for some `ctxt.Done()`

Reply:
Good question. This code converts the socket `Read` call to a channel. Ideally, all signals would be managed in the select statement below. But there is no channel API for the websocket read. A godoc explanation has been added in 8273d82d

The loop below does monitor the `ctx.Done()` channel